File: contest/ICPC/2014/欧洲/NEERC东北欧/std/problems/hidden/hidden_al.cpp
/******************************************************************/
/* ACM ICPC 2014-2015                                             */
/* Northeastern European Regional Contest                         */
/* St Petersburg - Barnaul - Tbilisi - Tashkent, December 7, 2014 */
/******************************************************************/
/* Problem H. Hidden Maze                                         */
/*                                                                */
/* Original idea         Alexander Kaluzhin                       */
/* Problem statement     Andrei Lopatin                           */
/* Test set              Andrei Lopatin                           */
/******************************************************************/
/* Solution                                                       */
/*                                                                */
/* Author                Andrei Lopatin                           */
/******************************************************************/

#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <algorithm>
#include <vector>

using namespace std;

typedef long long int64;
typedef pair <int, int> pii;

#define forn(i, n) for (int i = 0; i < (int) n; i++)
#define seta(a, b) memset(a, b, sizeof(a))
#define pb push_back
#define all(a) (a).begin(), (a).end()
#define last(a) a[a.size() - 1]
#define mp make_pair
#define fs first
#define sc second

const int NMAX = 30010;

int n, used[NMAX], h[NMAX], p[NMAX], d[NMAX], status[NMAX];
int64 ans, cnt;
vector<int> g[NMAX], w[NMAX];
vector<pii> edges;

struct Tstr {
	int n;
	vector<int> zn;

	Tstr() {
		Tstr(0);
	}

	Tstr(int n): n(n) {
		zn.resize(2 * n + 1);
		zn[n] = 1;
	}

	int& operator[](int idx) {
		return zn[idx + n];
	}

	int operator[](int idx) const {
		if (idx < -n || idx > n) return 0;
		return zn[idx + n];
	}
};

void add(Tstr& a, const Tstr& b, int sh) {
	for (int i = -b.n; i <= b.n; ++i) {
		a[i + sh] += b[i];
	}
}

void sub(Tstr& a, const Tstr& b, int sh) {
	for (int i = -b.n; i <= b.n; ++i)
		a[i + sh] -= b[i];
}

int64 mult(const Tstr& a, const Tstr& b, int sh) {
	int64 result = 0;
	for (int i = -b.n; i <= b.n; ++i)
		result += a[-(i + sh)] * b[i];
	return result;
}

Tstr str[NMAX];

void init(int v) {
	status[v] = 1;
	str[v] = Tstr(d[v]);
	used[v] = 1;
	forn(i, g[v].size())
		if (!used[g[v][i]]) {
			init(g[v][i]);
			add(str[v], str[g[v][i]], 1);
		}
}

void update(int v) {
	int now = 0, u = v;
	status[v] = 0;
	while (u != 0) {
		now += status[u];
		sub(str[p[u]], str[v], now + 1);
		add(str[p[u]], str[v], now - 1);
		u = p[u];
	}
	status[v] = -1;
}

int64 calc(int v) {
	int64 result = 0;
	int now = -status[v], u = v;
	while (u != 0) {
		now += status[u];
		result += mult(str[p[u]], str[v], now);
		result -= mult(str[u], str[v], now + status[u]);
		u = p[u];
	}
	return result;
}

void dfs(int v, int vv, int hh) {
	h[v] = hh;
	p[v] = vv;
	d[v] = 0;
	used[v] = 1;
	forn(i, g[v].size())
		if (!used[g[v][i]]) {
			dfs(g[v][i], v, hh + 1);
			d[v] = max(d[v], 1 + d[g[v][i]]);
		}
}

int main() {
	freopen("hidden.in", "rt", stdin);
	freopen("hidden.out", "wt", stdout);
	cin >> n;
	forn(i, n - 1) {
		int x, y, z;
		scanf("%d%d%d", &x, &y, &z);
		x--, y--;
		g[x].pb(y);
		w[x].pb(z);
		g[y].pb(x);
		w[y].pb(z);
	}

	ans = 0, cnt = 0;
	seta(used, 0);
	dfs(0, 0, 0);
	seta(used, 0);
	init(0);

	forn(i, n)
		forn(j, g[i].size())
			if (h[i] > h[g[i][j]])
				edges.pb(mp(w[i][j], i));
	sort(all(edges));

	forn(i, edges.size()) {
		int64 ept = calc(edges[i].sc);
		ans += ept * 1LL * edges[i].fs;
		cnt += ept;
		update(edges[i].sc);
	}

	printf("%.10lf\n", ans * 1. / cnt);

	return 0;
}

File: contest/ICPC/2014/欧洲/NEERC东北欧/std/problems/hidden/hidden_al.java
/******************************************************************/
/* ACM ICPC 2014-2015                                             */
/* Northeastern European Regional Contest                         */
/* St Petersburg - Barnaul - Tbilisi - Tashkent, December 7, 2014 */
/******************************************************************/
/* Problem H. Hidden Maze                                         */
/*                                                                */
/* Original idea         Alexander Kaluzhin                       */
/* Problem statement     Andrei Lopatin                           */
/* Test set              Andrei Lopatin                           */
/******************************************************************/
/* Solution                                                       */
/*                                                                */
/* Author                Andrei Lopatin                           */
/******************************************************************/

import java.io.*;
import java.util.*;

/**
 * Reference solution for NEERC'2014 Problem H: Hidden Maze
 * This solution checks correctness of the input.
 *
 * @author Andrei Lopatin
 */
public class hidden_al {
    private static final String INPUT = "hidden.in";
    private static final String OUTPUT = "hidden.out";

    public static void main(String[] args) throws Exception {
        new hidden_al().go();
    }

    void go() throws IOException {
        long time = System.currentTimeMillis();
        try (Scanner in = new Scanner(new File(INPUT))) {
            in.useLocale(Locale.US);
            read(in);
        }
        solve();
        try (PrintWriter out = new PrintWriter(OUTPUT)) {
            write(out);
        }
        System.out.printf("Done in %d ms%n", System.currentTimeMillis() - time);
    }

    static final int MAXN = 30000;
    static final int MAXC = (int)1e6;


    static class Edge implements Comparable<Edge> {
	int b, e, v;
	Edge(int _b, int _e, int _v) {
	    b = _b;
	    e = _e;
	    v = _v;
	}
	public int compareTo(Edge e2) {
	    return v - e2.v;
	}
    }

    Edge [] el;

    int n;
    ArrayList <Integer> e[];
    ArrayList <Integer> value[];

    void read(Scanner in) {
	n = in.nextInt();
	in.nextLine();
	assert 1 <= n && n <= MAXN;
	allocate();

	for (int i = 1; i < n; i++) {
	    int a = in.nextInt(), b = in.nextInt(), c = in.nextInt();
	    in.nextLine();
	    assert 1 <= a && a <= n && 1 <= b && b <= n && 1 <= c && c <= MAXC;
	    el[i - 1] = new Edge(a, b, c);
	}
    }

    void allocate() {
        e = new ArrayList[n + 1];
        value = new ArrayList[n + 1];
        el = new Edge[n - 1];
    }

    boolean[] f;
    int MaxD;
    int[][] S;
    int[] u, d, cu;

    void v_dfs(int v, int d) {
	if (d > MaxD) {
	    throw new AssertionError("Tree depth more than " + MaxD);
	}
	f[v] = true;
	for (int i = 0; i < e[v].size (); i++) {
	    int w = e[v].get(i);
	    if (!f[w]) {
		v_dfs(w, d + 1);
	    }
	}
    }

    void validate() {
	f = new boolean [n + 1];
	MaxD = 2 * (int)Math.sqrt(n) + 10;
	v_dfs(1, 0);
	for (int i = 1; i <= n; i++) {
	    if (!f[i]) {
		throw new AssertionError("Vertex " + i + " not reachable");
	    }
	}
    }

    int dfs_init(int v) {
	int cnt = 1, subd = 0;
	f[v] = true;
	for (int i = 0; i < e[v].size (); i++) {
	    int w = e[v].get(i);
	    if (!f[w]) {
		u[w] = v;
		cu[w] = value[v].get(i);
		subd = Math.max(subd, dfs_init(w) + 1);
		for (int j = 0; j < MaxD; j++) {
		    S[v][MaxD + j + 1] += S[w][MaxD + j];
		}
	    }
	}
	S[v][MaxD + 0] = 1;
	d[v] = subd;
	return subd;
    }

    void build_graph() {
	for (int i = 1; i <= n; i++) {
	    e[i] = new ArrayList <Integer>();
	    value[i] = new ArrayList <Integer>();
	}
	for (int i = 0; i < el.length; i++) {
	    e[el[i].b].add(el[i].e);
	    e[el[i].e].add(el[i].b);
	    value[el[i].b].add(i + 1);
	    value[el[i].e].add(i + 1);
	}
    }

    long sum, cnt;

    void solve() {
	Arrays.sort(el);

	build_graph();
	validate();

	S = new int[n + 1][MaxD * 2 + 1];
	u = new int[n + 1];
	d = new int[n + 1];
	cu = new int[n + 1];

	f = new boolean [n + 1];
	dfs_init(1);

	for (int i = 0; i < el.length; i++) {
	    int a = el[i].b, b = el[i].e;
	    if (u[a] == b) {
		int t = a;
		a = b;
		b = t;
	    }
	    // a - upper vertex, b - lower vertex
	    assert cu[b] == i + 1 && u[b] == a;
	    int ca = a, cb = b, cd = 0, ccnt = 0;
	    while (ca > 0) {

		int lb = Math.max(MaxD - d[ca], MaxD - d[b] - cd);
		int ub = Math.min(MaxD + d[ca], MaxD + d[b] - cd);

		int dd;
		if (cu[cb] > i) {
		    dd = 1;
		} else {
		    dd = -1;
		}

		for (int k = lb; k <= ub; k++) {
//		    System.out.println(b + " " + cb + " " + dd + " " + MaxD + " " + k + " " + S[ca][k] + " " + S[cb][k - dd] + " " + (2 * MaxD - k - cd) + " " + S[b][2 * MaxD - k - cd]);
		    assert S[ca][k] >= S[cb][k - dd];
		    ccnt += (S[ca][k] - S[cb][k - dd]) * S[b][2 * MaxD - k - cd];
		}

		if (cu[ca] > i + 1) {
		    ++cd;
		} else {
		    --cd;
		}
		cb = ca;
		ca = u[cb];
	    }
	    sum += (long)ccnt * el[i].v;
	    cnt += ccnt;

	    ca = a; cb = b; cd = 1;
	    while (ca > 0) {
		int lb = Math.max(MaxD - d[ca], MaxD - d[b] + cd);
		int ub = Math.min(MaxD + d[ca], MaxD + d[b] + cd);
		for (int k = lb; k <= ub; k++) {
		    S[ca][k] -= S[b][k - cd];
		    assert S[ca][k] >= 0;
		    S[ca][k - 2] += S[b][k - cd];
		}
		if (cu[ca] > i + 1) {
		    ++cd;
		} else {
		    --cd;
		}
		cb = ca;
		ca = u[cb];
	    }
	}
	//System.out.println(sum + " " + cnt);
    }

    void write(PrintWriter out) {
	out.println(1.0 * sum / cnt);
	out.close();
    }

//----------------- just for validation ------------------

    /**
     * Strict scanner to verify 100% correspondence between input files and input file format specification.
     * It is a drop-in replacement for {@link java.util.Scanner} that could be added to a solution source
     * (cut-and-paste) without breaking its ability to work with {@link java.util.Scanner}.
     */
    public class Scanner implements Closeable {
        private final BufferedReader in;
        private String line = "";
        private int pos;
        private int lineNo;
        private boolean localeset;

        public Scanner(File source) throws FileNotFoundException {
            in = new BufferedReader(new FileReader(source));
            nextLine();
        }

        public void close() {
            assert line == null : "Extra data at the end of file";
            try {
                in.close();
            } catch (IOException e) {
                throw new AssertionError("Failed to close with " + e);
            }
        }

        public void nextLine() {
            assert line != null : "EOF";
            assert pos == line.length() : "Extra characters on line " + lineNo;
            try {
                line = in.readLine();
            } catch (IOException e) {
                throw new AssertionError("Failed to read line with " + e);
            }
            pos = 0;
            lineNo++;
        }

        public String next() {
            assert line != null : "EOF";
            assert line.length() > 0 : "Empty line " + lineNo;
            if (pos == 0)
                assert line.charAt(0) > ' ' : "Line " + lineNo + " starts with whitespace";
            else {
                assert pos < line.length() : "Line " + lineNo + " is over";
                assert line.charAt(pos) == ' ' : "Wrong whitespace on line " + lineNo;
                pos++;
                assert pos < line.length() : "Line " + lineNo + " is over";
                assert line.charAt(pos) > ' ' : "Line " + lineNo + " has double whitespace";
            }
            StringBuilder sb = new StringBuilder();
            while (pos < line.length() && line.charAt(pos) > ' ')
                sb.append(line.charAt(pos++));
            return sb.toString();
        }

        public int nextInt() {
            String s = next();
            assert s.length() == 1 || s.charAt(0) != '0' : "Extra leading zero in number " + s + " on line " + lineNo;
            assert s.charAt(0) != '+' : "Extra leading '+' in number " + s + " on line " + lineNo;
            try {
                return Integer.parseInt(s);
            } catch (NumberFormatException e) {
                throw new AssertionError("Malformed number " + s + " on line " + lineNo);
            }
        }

        public double nextDouble() {
            assert localeset : "Locale must be set with useLocale(Locale.US)";
            String s = next();
            assert s.length() == 1 || s.startsWith("0.") || s.charAt(0) != '0' : "Extra leading zero in number " + s + " on line " + lineNo;
            assert s.charAt(0) != '+' : "Extra leading '+' in number " + s + " on line " + lineNo;
            try {
                return Double.parseDouble(s);
            } catch (NumberFormatException e) {
                throw new AssertionError("Malformed number " + s + " on line " + lineNo);
            }
        }

        public void useLocale(Locale locale) {
            assert locale == Locale.US;
            localeset = true;
        }
    }
}

